File: python/remove_k_digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack: list[str] = list()
        for i in range(len(num)):
            while len(stack) > 0 and k > 0 and stack[-1] > num[i]:
                stack.pop()
                k -= 1
            stack.append(num[i])

        while k > 0 and len(stack) > 0:
            stack.pop()
            k -= 1

        p: int = 0
        for i in range(len(stack)):
            if stack[i] == "0":
                p += 1
            elif stack[i] != "0":
                break
        result: str = "".join(stack[p::])

        return "0" if len(result) == 0 else result


def test_removeKdigits_case_1():
    # arrange
    num: str = "1432219"
    k: int = 3
    expected: str = "1219"

    # act
    solution = Solution()
    actual = solution.removeKdigits(num, k)

    # assert
    assert expected == actual


def test_removeKdigits_case_2():
    # arrange
    num: str = "10200"
    k: int = 1
    expected: str = "200"

    # act
    solution = Solution()
    actual = solution.removeKdigits(num, k)

    # assert
    assert expected == actual


def test_removeKdigits_case_3():
    # arrange
    num: str = "10"
    k: int = 2
    expected: str = "0"

    # act
    solution = Solution()
    actual = solution.removeKdigits(num, k)

    # assert
    assert expected == actual


def test_removeKdigits_case_4():
    # arrange
    num: str = "9"
    k: int = 1
    expected: str = "0"

    # act
    solution = Solution()
    actual = solution.removeKdigits(num, k)

    # assert
    assert expected == actual


def test_removeKdigits_case_5():
    # arrange
    num: str = "10001"
    k: int = 4
    expected: str = "0"

    # act
    solution = Solution()
    actual = solution.removeKdigits(num, k)

    # assert
    assert expected == actual

File: rust/src/code/remove_k_digits.rs
struct Solution {}

impl Solution {
    pub fn remove_kdigits(num: String, mut k: i32) -> String {
        let mut stack: Vec<u8> = Vec::new();
        let num: Vec<u8> = num.chars().map(|x| x as u8).collect();
        for i in 0..num.len() {
            while stack.len() > 0 && k > 0 && stack[stack.len() - 1] > num[i] {
                stack.pop().unwrap();
                k -= 1;
            }
            stack.push(num[i]);
        }

        while stack.len() > 0 && k > 0 {
            stack.pop().unwrap();
            k -= 1;
        }

        let stack: Vec<char> = stack.iter().map(|x| *x as char).collect();
        let mut p: usize = 0;
        for i in 0..stack.len() {
            if stack[i] == '0' {
                p += 1;
            } else if stack[i] != '0' {
                break;
            }
        }

        let result: String = stack[p..].iter().collect();
        return if result.len() == 0 {
            "0".to_string()
        } else {
            result
        };
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_remove_kdigits_case_1() {
        // arrange
        let num: String = "1432219".to_string();
        let k: i32 = 3;
        let expected: String = "1219".to_string();

        // act
        let mut actual = Solution::remove_kdigits(num, k);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_remove_kdigits_case_2() {
        // arrange
        let num: String = "10200".to_string();
        let k: i32 = 1;
        let expected: String = "200".to_string();

        // act
        let mut actual = Solution::remove_kdigits(num, k);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_remove_kdigits_case_3() {
        // arrange
        let num: String = "10".to_string();
        let k: i32 = 2;
        let expected: String = "0".to_string();

        // act
        let mut actual = Solution::remove_kdigits(num, k);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_remove_kdigits_case_4() {
        // arrange
        let num: String = "9".to_string();
        let k: i32 = 1;
        let expected: String = "0".to_string();

        // act
        let mut actual = Solution::remove_kdigits(num, k);

        // assert
        assert_eq!(expected, actual);
    }
}
